constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/ConstraintSetParser.java

```diff
@@ -726,6 +737,114 @@ static void parseChain(int orientation, State state,
         }
     }
 
+    private static float toPix(State state, float dp){
+       return state.getDpToPixel().toPixels(dp);
+    }
+    /**
+     * Support parsing Chain in the following manner
+     * chainId : {
+     *      type:'hChain'  // or vChain
+     *      contains: ['id1', 'id2', 'id3' ]
+     *      contains: [['id', weight, marginL ,marginR], 'id2', 'id3' ]
+     *      start: ['parent', 'start',0],
+     *      end: ['parent', 'end',0],
+     *      top: ['parent', 'top',0],
+     *      bottom: ['parent', 'bottom',0],
+     *      style: 'spread'
+     * }
+
+     * @throws CLParsingException
+     */
+    static void parseChainType(String orientation,
+            State state,
+            String chainName,
+            LayoutVariables margins,
+            CLObject object) throws CLParsingException {
+
+        ChainReference chain = (orientation.charAt(0) == 'h')
+                ? state.horizontalChain() : state.verticalChain();
+        chain.setKey(chainName);
+
+        for (String params : object.names()) {
+            switch (params) {
+                case "contains":
+                    CLElement refs = object.get(params);
+                    if (!(refs instanceof CLArray) || ((CLArray) refs).size() < 1) {
+                        System.err.println(
+                                chainName + " contains should be an array \"" + refs.content()
+                                        + "\"");
+                        return;
+                    }
+                    for (int i = 0; i < ((CLArray) refs).size(); i++) {
+                        CLElement chainElement = ((CLArray) refs).get(i);
+                        Utils.log("================ " + i + " " + chainElement.getClass() + "  "
+                                + chainElement.content());
+                        if (chainElement instanceof CLArray) {
+                            CLArray array = (CLArray) chainElement;
+                            if (array.size() == 1) {
+                                chain.add(array.get(0).content());
+                            } else {
+                                String id = array.get(0).content();
+                                float weight = Float.NaN;
+                                float preMargin = Float.NaN;
+                                float postMargin = Float.NaN;
+
+                                if (array.size() >= 2) {
```

Comment:
nit: we wouldn't need this if statement since codes in this else block should have array.size() is more than one.

I assume we shouldn't have an empty array in this case, otherwise, there would be an issue with this line 
**String id = array.get(0).content();**

Reply:
[a] = array.size() == 1
[a, b]  = array.size() =2

if  array.size()  == 1 
weight = array.getFloat(1); would crash

therefor array.size() must be greater than or equal(>=)  to 2 for array.getFloat(1) to work

Reply:
rewrote